src/Multi_filtration/include/gudhi/One_critical_filtration.h

```diff
@@ -39,266 +42,477 @@ namespace Gudhi::multi_filtration {
+  // considered as -inf.
+  One_critical_filtration(int n) : Base(n, -T_inf) {};  // minus infinity by default
+  One_critical_filtration(int n, T value) : Base(n, value) {};
+  One_critical_filtration(std::initializer_list<T> init) : Base(init) {};
+  One_critical_filtration(const std::vector<T> &v) : Base(v) {};
+  One_critical_filtration(std::vector<T> &&v) : Base(std::move(v)) {};
   One_critical_filtration(typename std::vector<T>::iterator it_begin, typename std::vector<T>::iterator it_end)
-      : std::vector<T>(it_begin, it_end) {};
+      : Base(it_begin, it_end) {};
   One_critical_filtration(typename std::vector<T>::const_iterator it_begin,
                           typename std::vector<T>::const_iterator it_end)
-      : std::vector<T>(it_begin, it_end) {};
+      : Base(it_begin, it_end) {};
+
+  One_critical_filtration &operator=(const One_critical_filtration &a)
+  {
+    Base::operator=(a);
+    return *this;
+  }
+
+  // HERITAGE
 
   using std::vector<T>::operator[];
   using value_type = T;
-  using OneCritical = One_critical_filtration<T>;
-  template <typename value_type>
-  using Base = One_critical_filtration<value_type>;
+
+  // CONVERTERS
+
   operator std::vector<T> &() const { return *this; }
-  std::vector<T> get_vector() const { return static_cast<std::vector<T>>(*this); }
-  std::size_t num_parameters() const { return this->size(); }
-  // TODO : REMOVE SIZE ( confusing when kcritical)
-  // static std::size_t num_generators() {return 1;}
 
-  operator std::vector<T>() const { return static_cast<std::vector<T>>(*this); }
+  operator std::vector<T>() const { return static_cast<std::vector<T> >(*this); }
+
+  // needed in python
+  friend One_critical_filtration get_copy(const One_critical_filtration &f) { return f; }
+
+  // like numpy
+  template <typename U>
+  One_critical_filtration<U> as_type() const
+  {
+    One_critical_filtration<U> out;
+    out.reserve(Base::size());
+    for (std::size_t i = 0u; i < Base::size(); i++) out.push_back(static_cast<U>(Base::operator[](i)));
+    return out;
+  }
+
+  // ACCESS
+
+  std::size_t num_parameters() const { return Base::size(); }
 
-  inline const std::vector<T> &as_vector() const { return *this; }
+  constexpr static One_critical_filtration inf() { return {T_inf}; }
 
-  inline bool is_inf() const {
-    if (this->size() != 1) return false;
-    return (this->operator[](0) == T_inf);
+  constexpr static One_critical_filtration minus_inf() { return {-T_inf}; }
+
+  constexpr static One_critical_filtration nan()
+  {
+    if constexpr (std::numeric_limits<T>::has_quiet_NaN) {
+      return {std::numeric_limits<T>::quiet_NaN()};
+    } else {
+      return {};  // to differentiate it from 0, an empty filtration value can't do much anyway.
+    }
   }
-  inline bool is_minus_inf() const {
+
+  // DESCRIPTORS
+
+  bool is_inf() const
+  {
+    if (Base::size() != 1) return false;
+    return (Base::operator[](0) == T_inf);
+  }
```

Comment:
I think that `{inf,...,inf}` should be treated as `{inf}` as this is error-prone (and even modify the value with some pointer magic to keep the const ?).